src/main/java/uk/ac/cam/cl/dtg/isaac/api/managers/EventBookingManager.java

```diff
@@ -392,6 +399,101 @@ public EventBookingDTO requestBooking(final IsaacEventPageDTO event, final Regis
+                    // if the user has previously cancelled we should let them book again.
+                    reservation = this.bookingPersistenceManager.updateBookingStatus(event.getId(), user.getId(), reservingUser.getId(),
+                            BookingStatus.RESERVED, null);
+                } else {
+                    reservation = this.bookingPersistenceManager.createBooking(event.getId(), user.getId(), reservingUser.getId(),
+                            BookingStatus.RESERVED, null);
+                }
+                reservations.add(reservation);
+
+                try {
+                    // af599 TODO: Use the correct email template here.
+                    // af599 TODO: Content team action.
+                    emailManager.sendTemplatedEmailToUser(user,
+                            emailManager.getEmailTemplateDTO("email-event-booking-confirmed"),
+                            new ImmutableMap.Builder<String, Object>()
+                                    .put("contactUsURL", generateEventContactUsURL(event))
+                                    .put("authorizationLink", String.format("https://%s/account?authToken=%s",
+                                            propertiesLoader.getProperty(HOST_NAME), event.getIsaacGroupToken()))
+                                    .put("event.emailEventDetails", event.getEmailEventDetails() == null ? "" : event.getEmailEventDetails())
+                                    .put("event", event)
+                                    .build(),
+                            EmailType.SYSTEM,
+                            Collections.singletonList(generateEventICSFile(event, reservation)));
+
+                } catch (SegueDatabaseException | ContentManagerException e) {
+                    log.error(String.format("Unable to send welcome email (%s) to user (%s)",
+                            event.getId(), user.getEmail()), e);
+                }
+
+                // auto add them to the group and grant the owner permission
+                if (event.getIsaacGroupToken() != null && !event.getIsaacGroupToken().isEmpty()) {
+                    try {
+                        this.userAssociationManager.createAssociationWithToken(event.getIsaacGroupToken(), user);
+                    } catch (InvalidUserAssociationTokenException e) {
+                        log.error(String.format("Unable to auto add user (%s) using token (%s) as the token is invalid.",
+                                user.getEmail(), event.getIsaacGroupToken()));
+                    }
+                }
+            }
+            return reservations;
+
+        } catch (EventIsFullException | SegueDatabaseException e) {
+            for (EventBookingDTO reservation : reservations) {
+                try {
+                    bookingPersistenceManager.deleteBooking(event.getId(), reservation.getUserBooked().getId());
+                } catch (SegueDatabaseException f) {
+                    // Tough luck?
```

Comment:
Yeah - although with the events system we could use real database transactions if we wanted to push more logic down to the persistence manager.